File: 256. Paint House.cpp
/*
  Time complexity: O(N)
  Space complexity: O(1)  (modifying the input array)
*/

int minCost(vector<vector<int>>& costs) {
  int n = costs.size();

  for(int i=n-2;i>=0;i--){
    costs[i][0] = costs[i][0] + min(costs[i+1][1],costs[i+1][2]);
    costs[i][1] = costs[i][1] + min(costs[i+1][0],costs[i+1][2]);
    costs[i][2] = costs[i][2] + min(costs[i+1][0],costs[i+1][1]);
  }

  return min({costs[0][0], costs[0][1], costs[0][2]});
}

File: 256.PaintHouse.py
class Solution:
    def minCost(self, costs: List[List[int]]) -> int:
        
        dp = [0,0,0]
        
        for i in range(len(costs)):
            dp_0 = costs[i][0] + min(dp[1], dp[2])
            dp_1 = costs[i][1] + min(dp[0], dp[2])
            dp_2 = costs[i][2] + min(dp[0], dp[1])
            dp = [dp_0,dp_1, dp_2]
            
        return min(dp)
            
        # space : o(1)
        # time : o(n)